les/lespay/server/prioritypool.go

```diff
@@ -0,0 +1,503 @@
+		activeQueue:       prque.NewLazyQueue(activeSetIndex, activePriority, activeMaxPriority, clock, lazyQueueRefresh),
+		inactiveQueue:     prque.New(inactiveSetIndex),
+		minCap:            minCap,
+		activeBias:        activeBias,
+		capacityStepDiv:   capacityStepDiv,
+	}
+
+	ns.SubscribeField(pp.priorityField, func(node *enode.Node, state nodestate.Flags, oldValue, newValue interface{}) {
+		if newValue != nil {
+			c := &ppNodeInfo{
+				node:          node,
+				nodePriority:  newValue.(nodePriority),
+				activeIndex:   -1,
+				inactiveIndex: -1,
+			}
+			ns.SetFieldSub(node, pp.ppNodeInfoField, c)
+		} else {
+			ns.SetStateSub(node, nodestate.Flags{}, pp.ActiveFlag.Or(pp.InactiveFlag), 0)
+			if n, _ := pp.ns.GetField(node, pp.ppNodeInfoField).(*ppNodeInfo); n != nil {
+				pp.disconnectedNode(n)
+			}
+			ns.SetFieldSub(node, pp.CapacityField, nil)
+			ns.SetFieldSub(node, pp.ppNodeInfoField, nil)
+		}
+	})
+	ns.SubscribeState(pp.ActiveFlag.Or(pp.InactiveFlag), func(node *enode.Node, oldState, newState nodestate.Flags) {
+		if c, _ := pp.ns.GetField(node, pp.ppNodeInfoField).(*ppNodeInfo); c != nil {
+			if oldState.IsEmpty() {
+				pp.connectedNode(c)
+			}
+			if newState.IsEmpty() {
+				pp.disconnectedNode(c)
+			}
+		}
+	})
+	ns.SubscribeState(pp.updateFlag, func(node *enode.Node, oldState, newState nodestate.Flags) {
+		if !newState.IsEmpty() {
+			pp.updatePriority(node)
+		}
+	})
+	return pp
+}
+
+// RequestCapacity checks whether changing the capacity of a node to the given target
+// is possible (bias is applied in favor of other active nodes if the target is higher
+// than the current capacity).
+// If setCap is true then it also performs the change if possible. The function returns
+// the minimum priority needed to do the change and whether it is currently allowed.
+// If setCap and allowed are both true then the caller can assume that the change was
+// successful.
+// Note: priorityField should always be set before calling RequestCapacity. If setCap
+// is false then both InactiveFlag and ActiveFlag can be unset and they are not changed
+// by this function call either.
+// Note 2: this function should run inside a NodeStateMachine operation
+func (pp *PriorityPool) RequestCapacity(node *enode.Node, targetCap uint64, bias time.Duration, setCap bool) (minPriority int64, allowed bool) {
+	pp.lock.Lock()
+	pp.activeQueue.Refresh()
+	var updates []capUpdate
+	defer func() {
+		pp.lock.Unlock()
+		pp.updateFlags(updates)
+	}()
+
+	if targetCap < pp.minCap {
+		targetCap = pp.minCap
+	}
+	c, _ := pp.ns.GetField(node, pp.ppNodeInfoField).(*ppNodeInfo)
+	if c == nil {
+		log.Error("RequestCapacity called for unknown node", "id", node.ID())
+		return math.MaxInt64, false
+	}
+	var priority int64
+	if targetCap > c.capacity {
+		priority = c.nodePriority.EstMinPriority(pp.clock.Now()+mclock.AbsTime(bias), targetCap, false)
+	} else {
+		priority = c.nodePriority.Priority(pp.clock.Now(), targetCap)
+	}
```

Comment:
Shouldn't we reject the "non-priority" client to update the capacity in the first place?

Reply:
Indeed we should. I implemented it in the client pool because I think this rule belongs there (the priority pool is a more abstract layer, it does not even know about the difference between positive and negative balances).

Reply:
DONE